### PR TITLE
Recipe App

### PR DESCRIPTION
**Pull request by smunoz1988**
Hello reviewer!

Team: @Danfall369 / @smunoz1988  2 members

For this milestone, We made these changes:

**Core actions:**

- A login page.
- A registration page.
- A food list (with all CRUD implementation, except for 'update').
- A recipes list (with all CRUD implementation, except for 'update').
- A general shopping list view (all missing food for all current user recipes and total price).
- A list of all public recipes from other users with their names and total prices.

**Technical setup**

- Set up the repository on GitHub and use Gitflow.
- Set up Devise for authentication.
- Set authorization rules.

**Project Requirements:**

- Login page and registration page: built with Devise.

**Food list:**

- Displays a list of food added by the logged-in user as in the wireframe (display also quantity of a given food).
- Lead to a form that allows users to add new food.

**Recipes list:**

- Displays a list of recipes created by the logged-in user as in the wireframe.
- Lead to recipe details.
- If the user is the owner of the recipe, the user can delete it.

**Public recipe list:**

- Display a list of all public recipes.
- Lead to recipe details.

**Recipe details:**

- Display a toggle button that allows for a recipe to be made public or private.
- If the recipe is public or the user is the owner of the recipe, display the recipe details as in the wireframe.
- If the user is the owner of the recipe, lead to the form that allows the user to add new food.

**General shopping list view:**

- Show the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user).
- Count the total food items and total price of the missing food.
- Create a navigation menu that allows users to open all of the pages from the web.
- Write rspec tests
- Add style based on the layout of the wireframes provided. 

image

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  resources :home, only: [:index, :show]
+  resources :users do
+    resources :foods
+    resources :recipes do
+      member do
+        patch 'toggle'
+      end
+      resources :recipe_foods
+    end
+    resources :public_recipes
```
Comment (tugsem) [inline config/routes.rb]: - [ ] Please remove the endpoints you don't use as you did in the home path on line 3. This is for best practices.

File: app/views/recipes/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  <div class="container">
+    <%= button_to "Add recipe", new_user_recipe_path(user_id: @user.id), method: :get, class: "btn" %>
+    <br />
+    <% if @user.recipes.empty? %>
+      <p>No recipes added yet</p>
+    <% else %>
+      <% @user.recipes.each do |recipe| %>
+        <div class="recipe">
+          <p><%= link_to recipe.name, user_recipe_path(user_id: recipe.user_id, id: recipe.id), class: "recipe-name" %></p>
+          <p class="recipe-description"><%= recipe.description %></p>
```
Comment (tugsem) [inline app/views/recipes/index.html.erb]: - [ ] To match the given wireframe, please use flexbox to display the description and the recipe name on one line.

given|yours
---|---
image|image

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      <h1>Recipe On Rails</h1>
+    </div>  
+    <nav>
+      <ul>
+        <% if user_signed_in? %>
+          <li>Welcome, <%= current_user.name %></li>
+          <li><%= link_to "Foods", user_foods_path(current_user) %></li>
+          <li><%= link_to "Recipes", user_recipes_path(current_user) %></li>
+          <li><%= link_to "Public Recipes", user_public_recipes_path(current_user) %></li>
+          <li><%= link_to "Edit Profile", edit_user_registration_path %></li>
```
Comment (tugsem) [inline app/views/layouts/application.html.erb]: - [ ] Kindly add a nav link to redirect user to the general shopping list as per the requirement.

File: app/views/public_recipes/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    <h2>Public recipes</h2>
+
+
+  <div class= "public-recipe">
+    <% if @public_recipes.empty? %>
+      <p>No public recipes added yet</p>
+    <% else %>
+      <% @public_recipes.each do |recipe| %>
+        <div>
+          <h3><%= link_to recipe.name, user_recipe_path(user_id: recipe.user_id, id: recipe.id), class: "recipe-name" %></h3>
```
Comment (tugsem) [inline app/views/public_recipes/index.html.erb]: - [ ] In the public recipes, there will be many user's recipes. As per the requirement, if the recipe belongs to the author, there should be a remove button. Kindly implement this.

**Review by tugsem — CHANGES_REQUESTED** [review]
##  STATUS: REQUIRED CHANGES ♻️
Hello @smunoz1988  and @Danfall369 👋 , 

Good job so far! 👏 
There are some issues that you still need to work on to go to the next project but you are almost there!

Alt Text
## Highlights 🚀 
✔️database created according to the ERD
✔️app has both authentication and authorization gems used correctly
✔️layout matches the given wireframe
👏 Styling looks great!
## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/recipes/index.html.erb (hunk 2 above)
Comment (smunoz1988) [reply app/views/recipes/index.html.erb]: @tugsem  

changes done!

image

File: app/views/public_recipes/index.html.erb (hunk 4 above)
Comment (smunoz1988) [reply app/views/public_recipes/index.html.erb]: @tugsem 

I think you got confused. in the public recipes there's no delete button. 

image

image

but as requested, We added it! 

image

thanks for the suggestion.

File: app/views/layouts/application.html.erb (hunk 3 above)
Comment (smunoz1988) [reply app/views/layouts/application.html.erb]: @tugsem 

Done!

image

File: config/routes.rb (hunk 1 above)
Comment (smunoz1988) [reply config/routes.rb]: @tugsem 

Done! thanks for the suggestion!

**Comment by tugsem** [comment]
## STATUS: APPROVED 👍🎯  
Hello @smunoz1988  and @Danfall369 👋 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


## Highlights ⭐ 
✔️ changes implemented

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/public_recipes/index.html.erb (hunk 4 above)
Comment (tugsem) [reply app/views/public_recipes/index.html.erb]: @smunoz1988 Hi, I know the wireframe doesn't have a button, I should have said to add it to the recipe details. Sorry for the misunderstanding. Anyways, it is nicely done! 🍾

**Pull request merged**